src/coreclr/jit/valuenum.cpp

```diff
@@ -8589,12 +8596,18 @@ static bool GetStaticFieldSeqAndAddress(ValueNumStore* vnStore, GenTree* tree, s
     }
 
     // Base address is expected to be static field's address
-    if ((tree->IsCnsIntOrI()) && (tree->AsIntCon()->gtFieldSeq != nullptr) &&
-        (tree->AsIntCon()->gtFieldSeq->GetKind() == FieldSeq::FieldKind::SimpleStaticKnownAddress))
+    ValueNum treeVN = tree->gtVNPair.GetLiberal();
+    if (tree->gtVNPair.BothEqual() && vnStore->IsVNConstant(treeVN) && vnStore->IsVNHandle(treeVN) &&
+        (vnStore->GetHandleFlags(treeVN) == GTF_ICON_STATIC_HDL))
     {
-        *pFseq      = tree->AsIntCon()->gtFieldSeq;
-        *byteOffset = tree->AsIntCon()->IconValue() + val - tree->AsIntCon()->gtFieldSeq->GetOffset();
-        return true;
+        ssize_t   icon   = vnStore->CoercedConstantValue<ssize_t>(treeVN);
+        FieldSeq* fldSeq = vnStore->GetFieldSeqFromAddress(icon);
+        if (fldSeq != nullptr)
+        {
+            *pFseq      = fldSeq;
+            *byteOffset = icon + val - fldSeq->GetOffset();
```

Comment:
Here we will now always have `icon == fldSeq->GetOffset()`.

Reply:
Changed to `*byteOffset = val`